### PR TITLE
issue: 2816687 Fix dynamic key/certificate generation

### PR DESCRIPTION
**Pull request by alexbriskin**

### CODE REVIEW

File: src/tls.cpp [hunk 1]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+
+static inline EVP_PKEY* generate_EC_pkey_with_NID(int nid=NID_secp384r1)
+{
+    bool result;
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    result = ((ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL)) != NULL) &&
+             (EVP_PKEY_keygen_init(ctx) == 1) &&
+             (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, nid) > 0) &&
+             (EVP_PKEY_keygen(ctx, &pkey) == 1);
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: We probably need to call EVP_PKEY_CTX_free on ctx in case of an error in one of the function following EVP_PKEY_CTX_new_id. We probably also need to free ctx when done using it, which maybe not in this function.

File: src/tls.cpp [hunk 2]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+    RSA_generate_key_ex(rsa, 2048, bignum, nullptr) == 1 &&
+    EVP_PKEY_assign(pkey,EVP_PKEY_RSA, rsa) == 1;
+
+    BN_free(bignum);
+
+    if (result) {
+      return pkey;
+    } else {
+      RSA_free(rsa);
+      EVP_PKEY_free(pkey);
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: What about bignum ?

File: src/tls.cpp [hunk 3]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+	((bignum = BN_new()) != nullptr) &&
+	((rsa = RSA_new()) != nullptr) &&
+	BN_set_word(bignum, RSA_F4) == 1 &&
+    RSA_generate_key_ex(rsa, 2048, bignum, nullptr) == 1 &&
+    EVP_PKEY_assign(pkey,EVP_PKEY_RSA, rsa) == 1;
+
+    BN_free(bignum);
+
+    if (result) {
+      return pkey;
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: What should we do with rsa? When will it be freed?

File: src/tls.cpp [hunk 4]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+    } else {
+      RSA_free(rsa);
+      EVP_PKEY_free(pkey);
+      return nullptr;
+    }
+}
+
+#define X509_add_feild(feild, value) \
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: Typo: ..field
Also, please consider a name that do not confuse with the library namings. Something like X509_ADD_FIELD.

File: src/tls.cpp [hunk 5]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+      (X509_add_feild("O", (unsigned char *)"Nvidia") == 1) &&
+      (X509_add_feild("CN", (unsigned char *)"localhost") == 1) &&
+      (X509_set_issuer_name(x509, name) == 1) &&
+      (X509_sign(x509, pkey, EVP_sha256()) != 0);
+
+    if (result) {
+      return x509;
+    } else {
+      X509_free(x509);
+      return NULL;
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: better use nullptr

File: src/tls.cpp [hunk 6]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+
+static inline bool add_ec_sec384r1_key_and_certificate(SSL_CTX *ctx)
+{
+  EVP_PKEY *pkey;
+  X509 *x509;
+
+  return ((pkey = generate_EC_pkey_with_NID()) != NULL) &&
+         ((x509 = generate_self_signed_x509_with_key(pkey)) != NULL) &&
+         add_key_and_certificate(ctx, pkey, x509);
+}
```
Comment (AlexanderGrissik) [inline src/tls.cpp:376]: Please use the same alignment style which is currently used in this file.

File: src/tls.cpp [hunk 7]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+#undef X509_add_feild
+
+static inline bool add_key_and_certificate(SSL_CTX *ctx,
+                                           EVP_PKEY *pkey,
+                                           X509 *x509)
+{
+  if ((SSL_CTX_use_certificate(ctx, x509) == 1)
+   && (SSL_CTX_use_PrivateKey(ctx, pkey) == 1)) {
+    return true;
+  } else {
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: You don't need else since you do return  in the previous statement.

File: src/tls.cpp [hunk 8]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+      ((name = X509_get_subject_name(x509)) != NULL) &&
+      (X509_add_feild("C", (unsigned char *)"IL") == 1) &&
+      (X509_add_feild("O", (unsigned char *)"Nvidia") == 1) &&
+      (X509_add_feild("CN", (unsigned char *)"localhost") == 1) &&
+      (X509_set_issuer_name(x509, name) == 1) &&
+      (X509_sign(x509, pkey, EVP_sha256()) != 0);
+
+    if (result) {
+      return x509;
+    } else {
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: You don't need else since you do return in the previous statement.

File: src/tls.cpp [hunk 9]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+      log_err("unable to use key");
+      return false;
+    }
+    return true;
+}
+
+static inline EVP_PKEY* generate_EC_pkey_with_NID(int nid=NID_secp384r1)
+{
+    bool result;
+    EVP_PKEY *pkey = NULL;
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: Better use nullptr

File: src/tls.cpp [hunk 10]
```diff
@@ -210,6 +85,7 @@ static int wait_for_single_socket(int fd, int which) {
     return ret;
 }
 
+static inline EVP_PKEY* generate_EC_pkey_with_NID(int nid);
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: Please put this together with previous pre declarations.

File: src/tls.cpp [hunk 11]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
     return SSL_read((SSL *)handle, buf, num);
 }
 
+static inline bool add_key_and_certificate(SSL_CTX *ctx, int keytype,
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: Is this method used somewhere?

File: src/tls.cpp [hunk 12]
```diff
@@ -26,140 +27,11 @@
  * OF SUCH DAMAGE.
  */
 
-#include "common.h"
+#include "common.h" // FIXME Circlular include/dependency?
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: You can try fix that by putting inside common.h:
extern tls_read(...)
extern tls_write(...)
and removing include tls from common.h

Comment (ChrisCoe) [inline src/tls.cpp]: Standard notation for this in codebase is `TODO` instead of `FIXME`.

File: src/tls.cpp [hunk 13]
```diff
@@ -360,6 +235,119 @@ int tls_read(void *handle, void *buf, int num) {
...
+  EVP_PKEY *pkey{nullptr};
+  BIGNUM *bignum{nullptr};
+
+  bool result =
+	((pkey = EVP_PKEY_new()) != nullptr) &&
+	((bignum = BN_new()) != nullptr) &&
+	((rsa = RSA_new()) != nullptr) &&
+	BN_set_word(bignum, RSA_F4) == 1 &&
+    RSA_generate_key_ex(rsa, 2048, bignum, nullptr) == 1 &&
+    EVP_PKEY_assign(pkey,EVP_PKEY_RSA, rsa) == 1;
```
Comment (ChrisCoe) [inline src/tls.cpp]: Align whitespace as done in lines 258-261.

File: src/tls.cpp [hunk 14]
```diff
@@ -26,140 +27,11 @@
  * OF SUCH DAMAGE.
  */
 
-#include "common.h"
+#include "common.h" // FIXME Circlular include/dependency?
 #include "tls.h"
 
 #if defined(DEFINED_TLS)
-
-static unsigned char cert_der[] = {
```
Comment (AlexanderGrissik) [inline src/tls.cpp:34]: Are we sure we want a generated cipher instead of a predefined one? Can it cause different performance results with different generations?

**Review by AlexanderGrissik — COMMENTED** [review]
Please see my comments

File: src/tls.cpp (hunk 12 above)
Comment (alexbriskin) [reply src/tls.cpp]: Aperanetely this is needed for the template initialization.
I won't address this in this PR.

File: src/tls.cpp (hunk 13 above)
Comment (alexbriskin) [reply src/tls.cpp]: Fixed

File: src/tls.cpp (hunk 9 above)
Comment (alexbriskin) [reply src/tls.cpp]: Done

File: src/tls.cpp (hunk 6 above)
Comment (alexbriskin) [reply src/tls.cpp:376]: OK

File: src/tls.cpp (hunk 8 above)
Comment (alexbriskin) [reply src/tls.cpp]: OK

File: src/tls.cpp (hunk 14 above)
Comment (alexbriskin) [reply src/tls.cpp:34]: It should only influence the initialization phase.

File: src/tls.cpp (hunk 12 above)
Comment (alexbriskin) [reply src/tls.cpp]: I decided not addressing the issue in this PR

File: src/tls.cpp (hunk 1 above)
Comment (alexbriskin) [reply src/tls.cpp]: Yes, you are right.

File: src/tls.cpp (hunk 2 above)
Comment (alexbriskin) [reply src/tls.cpp]: `bignum` unconditionally freed above

File: src/tls.cpp (hunk 3 above)
Comment (alexbriskin) [reply src/tls.cpp]: According to the EVP_PKEY_assign manual:
> however these use the supplied key internally without incrementing its reference count, such that key will be freed when the parent pkey is freed.

If I am reading this correctly pkey "owns" the rsa instance and is responsible for freeing it.

File: src/tls.cpp (hunk 4 above)
Comment (alexbriskin) [reply src/tls.cpp]: OK

File: src/tls.cpp (hunk 5 above)
Comment (alexbriskin) [reply src/tls.cpp]: Done

File: src/tls.cpp (hunk 11 above)
Comment (alexbriskin) [reply src/tls.cpp]: Yes, by `add_rsa2048_key_and_certificate` and `add_ec_sec384r1_key_and_certificate`

File: src/tls.cpp (hunk 10 above)
Comment (alexbriskin) [reply src/tls.cpp]: Removed the declaration as it was unnecessary.

File: src/tls.cpp (hunk 7 above)
Comment (alexbriskin) [reply src/tls.cpp]: OK

File: src/tls.cpp (hunk 11 above)
Comment (AlexanderGrissik) [reply src/tls.cpp]: You have two definitions of this function:
1. add_key_and_certificate(SSL_CTX *ctx, int keytype,unsigned char *cert, size_t cert_size,unsigned char *key, size_t key_size)
2. add_key_and_certificate(SSL_CTX *ctx, EVP_PKEY *pkey, X509 *x509)
The second one is used by add_rsa2048_key_and_certificate and add_ec_sec384r1_key_and_certificate.
The first one seems unused.

File: src/tls.cpp (hunk 12 above)
Comment (AlexanderGrissik) [reply src/tls.cpp]: Then lets rephrase it to: Caution: circular include dependency.

File: src/tls.cpp (hunk 14 above)
Comment (AlexanderGrissik) [reply src/tls.cpp:34]: Yes i understand that. What i mean is whether it may cause unstable performance results because between one execution of sockperf to the other execution it used a different actual key.

Comment (AlexanderGrissik) [reply src/tls.cpp:34]: I would suggest here, to prevent performance instability, that instead of auto generating different key each time. To add an option to supply sockperf with the needed key. Similar to how its done by nginx. It is also more suitable for debugging when you know what to expect.

File: src/tls.cpp [hunk 15]
```diff
@@ -360,6 +234,121 @@ int tls_read(void *handle, void *buf, int num) {
...
+    if (result) {
+        return pkey;
+    }
+
+    RSA_free(rsa);
+    EVP_PKEY_free(pkey);
+    return nullptr;
+}
+
+#define X509_add_field(field, value) \
```
Comment (AlexanderGrissik) [inline src/tls.cpp]: Please use big letter notation for macro

**Review by AlexanderGrissik — APPROVED** [review]

**Comment by igor-ivanov** [comment]
Moved to main branch as https://github.com/Mellanox/sockperf/commit/bd513323652eb28722f8690b1d27171d039f8a13
